lib/benchmarker.rb

```diff
@@ -0,0 +1,218 @@
+$LOAD_PATH.unshift File.expand_path('../', __FILE__)
+require 'thin'
+require 'benchmark'
+require 'benchmark/ips'
+require 'memory_profiler'
+require 'semian'
+require 'semian/net_http'
+require 'toxiproxy'
+require 'yaml'
+require 'byebug'
+
+class SemianConfig
+  CONFIG_FILE = File.expand_path('../../test/config/hosts.yml', __FILE__)
+
+  class << self
+    def [](service)
+      all.fetch(service)
+    end
+
+    def all
+      @entries ||= YAML.load_file(CONFIG_FILE)
+    end
+  end
+end
+
+class RackServer
+  def self.call(env)
+    response_code = env['REQUEST_URI'].delete("/")
+    response_code = '200' if response_code == ""
+    [response_code, {'Content-Type' => 'text/html'}, ['Success']]
+  end
+end
+
+
+class Benchmarker
+  DEFAULT_SEMIAN_OPTIONS = {
+    tickets: 3,
+    success_threshold: 1,
+    error_threshold: 3,
+    error_timeout: 10,
+  }.freeze
+
+  DEFAULT_SEMIAN_CONFIGURATION = proc do |host, port|
+    next nil if host == SemianConfig['toxiproxy_upstream_host'] && port == SemianConfig['toxiproxy_upstream_port'] # disable if toxiproxy
+    DEFAULT_SEMIAN_OPTIONS.merge(name: "#{host}_#{port}")
+  end
+
+  def initialize(number_of_resources)
+    run_benchmark_resources(number_of_resources)
+  end
+
+  def run_benchmark_resources(number_of_resources)
+    run_ips
+
+    # [50000].each do |nb_resoures|
+    #   lru_resource(nb_resoures)
+    # end
+    # execution_time = Benchmark.measure { lru_resource(number_of_resources) }
+    # puts "Execution time = #{execution_time}"
+  end
+
+  def run_ips
+    Benchmark.ips do |x|
+      [500, 1000, 2500, 5000, 10000].each do |nb_resoures|
+        x.report("#{nb_resoures} resources ") { lru_resource(nb_resoures) }
+      end
+      x.compare!
+    end
+  end
+
+  def lru_resource(number_of_resources)
+    @request_time = []
+    reset_semian_resource
+    with_semian_configuration do
+      with_server(ports: [3000], reset_semian_state: false) do |host, port|
+        number_of_resources.times do |i|
+          create_request(host, port, i)
+        end
+      end
+    end
+    # puts "Average = #{@request_time.reduce(:+) / @request_time.size.to_f}"
+  end
+
+  def create_request(host, port, i)
+    random = rand(1...100)
+    average_time = average do
+      if random >= 0 && random <= 50
```

Comment:
This allows us to have a `success` or `failure` request, this make sure the data is not always a `success` or a `failure`